README.rst
==============
SockJS-Twisted
==============

A simple library for adding SockJS support to your twisted application.

Status
======

SockJS-Twisted passes all `SockJS-Protocol v0.3.3 <https://github.com/sockjs/sockjs-protocol>`_ tests,
and all `SockJS-Client qunit <https://github.com/sockjs/sockjs-client>`_ tests. It has been used in
production environments, and should be free of any critical bugs.

Usage
=====

Use ``txsockjs.factory.SockJSFactory`` to wrap your factories. That's it!

.. code-block:: python
    
    from twisted.internet import reactor
    from twisted.internet.protocol import Factory, Protocol
    from txsockjs.factory import SockJSFactory

    class HelloProtocol(Protocol):
        def connectionMade(self):
            self.transport.write('hello')
            self.transport.write('how are you?')

        def dataReceived(self, data):
            print data

    reactor.listenTCP(8080, SockJSFactory(Factory.forProtocol(HelloProtocol)))
    reactor.run()

There is nothing else to it, no special setup involved.

Do you want a secure connection? Use ``listenSSL()`` instead of ``listenTCP()``.

Advanced Usage
==============

For those who want to host multiple SockJS services off of one port,
``txsockjs.factory.SockJSMultiFactory`` is designed to handle routing for you.

.. code-block:: python

    from twisted.internet import reactor
    from twisted.internet.protocol import Factory, Protocol
    from txsockjs.factory import SockJSMultiFactory
    from txsockjs.utils import broadcast

    class EchoProtocol(Protocol):
        def dataReceived(self, data):
            self.transport.write(data)

    class ChatProtocol(Protocol):
        def connectionMade(self):
            if not hasattr(self.factory, "transports"):
                self.factory.transports = set()
            self.factory.transports.add(self.transport)

        def dataReceived(self, data):
            broadcast(data, self.factory.transports)

        def connectionLost(self, reason):
            self.factory.transports.remove(self.transport)

    f = SockJSMultiFactory()
    f.addFactory(Factory.forProtocol(EchoProtocol), "echo")
    f.addFactory(Factory.forProtocol(ChatProtocol), "chat")

    reactor.listenTCP(8080, f)
    reactor.run()

http://localhost:8080/echo and http://localhost:8080/chat will give you access
to your EchoFactory and ChatFactory.

Integration With Websites
=========================

It is possible to offer static resources, dynamic pages, and SockJS endpoints off of
a single port by using ``txsockjs.factory.SockJSResource``.

.. code-block:: python

    from twisted.internet import reactor
    from twisted.internet.protocol import Factory, Protocol
    from twisted.web import resource, server
    from txsockjs.factory import SockJSResource

    # EchoProtocol and ChatProtocol defined above

    root = resource.Resource()
    root.putChild("echo", SockJSResource(Factory.forProtocol(EchoProtocol)))
    root.putChild("chat", SockJSResource(Factory.forProtocol(ChatProtocol)))
    site = server.Site(root)

    reactor.listenTCP(8080, site)
    reactor.run()

Multiplexing [Experimental]
===========================

SockJS-Twisted also has built-in support for multiplexing. See the
`Websocket-Multiplex <https://github.com/sockjs/websocket-multiplex>`_ library
for how to integrate multiplexing client side.

.. code-block:: python

    from twisted.internet import reactor
    from twisted.internet.protocol import Factory, Protocol
    from twisted.web import resource, server
    from txsockjs.multiplex import SockJSMultiplexResource

    multiplex = SockJSMultiplexResource()
    multiplex.addFactory("echo", Factory.forProtocol(EchoProtocol))
    multiplex.addFactory("chat", Factory.forProtocol(ChatProtocol))

    root = resource.Resource()
    root.putChild("multiplex", multiplex)
    site = server.Site(root)

    reactor.listenTCP(8080, site)
    reactor.run()

Single factory? Multifactory? Resource? Multiplexing? What's the difference?
============================================================================

+-------------------------+--------------------+----------------------------------+--------------------------+
| Type                    | Factories per port | Allows mixing native web content | Factories per connection |
+=========================+====================+==================================+==========================+
| SockJSFactory           | Single             | No                               | Single                   |
+-------------------------+--------------------+----------------------------------+--------------------------+
| SockJSMultiFactory      | Multiple           | No                               | Single                   |
+-------------------------+--------------------+----------------------------------+--------------------------+
| SockJSResource          | Multiple           | Yes                              | Single                   |
+-------------------------+--------------------+----------------------------------+--------------------------+
| SockJSMultiplexResource | Multiple           | Yes                              | Multiple                 |
+-------------------------+--------------------+----------------------------------+--------------------------+

``SockJSFactory`` is recommended for use in non-web (HTTP) applications to allow
native web connections. For instance, an IRC server. There can only be one factory
listening on a port using this method. The SockJS endpoint uses this internally.

``SockJSMultiFactory`` is recommended for use in non-web (HTTP) applications with
multiple services. This allows multiple factories to listen on a single port.

``SockJSResource`` is recommended for use in HTTP based applications, like webservers.

``SockJSMultiplexResource`` is recommended for pubsub applications, where each connection
needs to talk to multiple factories. Overriding the subscribe method allows for dynamic
factory creation if you don't know what is needed server-side ahead of time.

Endpoints
=========

For integration with pre-existing libraries or programs, it is possible use sockjs
as an endpoint in the form ``sockjs:tcp\:9090\:interface\=0.0.0.0:encoding=utf8:websocket=false``.
You can pass any escaped endpoint to a sockjs endpoint to wrap it with txsockjs, and you can
specify any option for the SockJSFactory by specifying it as a keyword argument.
For more information, read the
`twisted documentation on endpoints <http://twistedmatrix.com/documents/current/core/howto/endpoints.html>`_.

.. code-block:: python

    from twisted.internet import reactor
    from twisted.internet.protocol import Factory, Protocol
    from twisted.internet.endpoints import serverFromString
    # Note that we don't have to import anything from txsockjs

    # HelloProtocol defined above

    endpoint = serverFromString(reactor, "sockjs:tcp\:8080")
    endpoint.listen(Factory.forProtocol(HelloProtocol))
    reactor.run()

Options
=======

A dictionary of options can be passed into the factory to control SockJS behavior.

.. code-block:: python

    options = {
        'websocket': True,
        'cookie_needed': False,
        'heartbeat': 25,
        'timeout': 5,
        'streaming_limit': 128 * 1024,
        'encoding': 'cp1252', # Latin1
        'sockjs_url': 'https://d1fxtkz8shb9d2.cloudfront.net/sockjs-0.3.js',
        'proxy_header': None
    }
    SockJSFactory(factory_to_wrap, options)
    SockJSMultiFactory().addFactory(factory_to_wrap, prefix, options)
    SockJSResource(factory_to_wrap, options)
    SockJSMultiplexResource(options)

websocket :
    whether websockets are supported as a protocol. Useful for proxies or load balancers that don't support websockets.

cookie_needed :
    whether the JSESSIONID cookie is set. Results in less performant protocols being used, so don't require them unless your load balancer requires it.

heartbeat :
    how often a heartbeat message is sent to keep the connection open. Do not increase this unless you know what you are doing.

timeout :
    maximum delay between connections before the underlying protocol is disconnected

streaming_limit :
    how many bytes can be sent over a streaming protocol before it is cycled. Allows browser-side garbage collection to lower RAM usage.

encoding :
    All messages to and from txsockjs should be valid UTF-8. In the event that a message received by txsockjs is not UTF-8, fall back to this encoding.

sockjs_url :
    The url of the SockJS library to use in iframes. By default this is served over HTTPS and therefore shouldn't need changing.

proxy_header :
    The HTTP header to pull a proxied IP address out of. Leave as None to get the unproxied IP. **Do not change this unless you are behind a proxy you control.**

License
=======

SockJS-Twisted is (c) 2012 Christopher Gamble and is made available under the BSD license.


CHANGELOG.rst
========================
SockJS-Twisted Changelog
========================

1.2
===

**1.2.2**
 * Fix CVE-2014-4671
 * Fix numerous bugs in which unicode wasn't converted to UTF-8
 * Fix heartbeats not being sent on websocket transport
 * Add MIT license headers to enable using txsockjs in Debian packages

**1.2.1**
 * Fix broken setup.py

**1.2.0**
 * Add endpoint support

1.1
===

**1.1.1**
 * Add python 2.6 compatability
 * Fix heartbeats not being sent until data was written from wrapped protocol

**1.1.0**
 * Add proxy_header to expose proxied IPs
 * Remove pubsub from multiplexing as it was misleading

1.0
===

**1.0.0**
 * Refactor entire library to use twisted.web
 * Use t.w.websockets for websocket support instead of txWS
 * Add experimental multiplexing and pubsub functionality

0.1
===

**0.1.1**
 * Converts all text to UTF-8 (prevents crashing websockets in chrome)
 * Minor bug fixes

**0.1.0**
 * Initial release


txsockjs/factory.py
# Copyright (c) 2012, Christopher Gamble
# All rights reserved.
# 
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#    * Redistributions of source code must retain the above copyright
#      notice, this list of conditions and the following disclaimer.
#    * Redistributions in binary form must reproduce the above copyright
#      notice, this list of conditions and the following disclaimer in the
#      documentation and/or other materials provided with the distribution.
#    * Neither the name of the Christopher Gamble nor the names of its 
#      contributors may be used to endorse or promote products derived 
#      from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED 
# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
# IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
# INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
# LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
# OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
# OF THE POSSIBILITY OF SUCH DAMAGE.

from twisted.web import resource, server
from txsockjs.protocols.base import Stub
from txsockjs.protocols.eventsource import EventSource
from txsockjs.protocols.htmlfile import HTMLFile
from txsockjs.protocols.jsonp import JSONP, JSONPSend
from txsockjs.protocols.static import Info, IFrame
from txsockjs.protocols.websocket import RawWebSocket, WebSocket
from txsockjs.protocols.xhr import XHR, XHRSend, XHRStream

class SockJSFactory(server.Site):
    def __init__(self, factory, options = None):
        server.Site.__init__(self, SockJSResource(factory, options))

class SockJSMultiFactory(server.Site):
    def __init__(self):
        server.Site.__init__(self, resource.Resource())
    
    def addFactory(self, factory, prefix, options = None):
        self.resource.putChild(prefix, SockJSResource(factory, options))

class SockJSResource(resource.Resource):
    def __init__(self, factory, options = None):
        resource.Resource.__init__(self)
        self._factory = factory
        self._sessions = {}
        self._options = {
            'websocket': True,
            'cookie_needed': False,
            'heartbeat': 25,
            'timeout': 5,
            'streaming_limit': 128 * 1024,
            'encoding': 'cp1252', #Latin1
            'sockjs_url': 'https://d1fxtkz8shb9d2.cloudfront.net/sockjs-0.3.js',
            'proxy_header': None
        }
        if options is not None:
            self._options.update(options)
        # Just in case somebody wants to mess with these
        self._methods = {
            'xhr': XHR,
            'xhr_send': XHRSend,
            'xhr_streaming': XHRStream,
            'eventsource': EventSource,
            'htmlfile': HTMLFile,
            'jsonp': JSONP,
            'jsonp_send': JSONPSend,
        }
        self._writeMethods = ('xhr_send','jsonp_send')
        # Static Resources
        self.putChild("info",Info())
        self.putChild("iframe.html",IFrame())
        self.putChild("websocket",RawWebSocket())
        # Since it's constant, we can declare the websocket handler up here
        self._websocket = WebSocket()
        self._websocket.parent = self
    
    def getChild(self, name, request):
        # Check if it is the greeting url
        if not name and not request.postpath:
            return self
        # Hacks to resove the iframe even when people are dumb
        if len(name) > 10 and name[:6] == "iframe" and name[-5:] == ".html":
            return self.children["iframe.html"]
        # Sessions must have 3 parts, name is already the first. Also, no periods in the loadbalancer
        if len(request.postpath) != 2 or "." in name or not name:
            return resource.NoResource("No such child resource.")
        # Extract session & request type. Discard load balancer
        session, name = request.postpath
        # No periods in the session
        if "." in session or not session:
            return resource.NoResource("No such child resource.")
        # Websockets are a special case
        if name == "websocket":
            return self._websocket
        # Reject invalid methods
        if name not in self._methods:
            return resource.NoResource("No such child resource.")
        # Reject writes to invalid sessions, unless just checking options
        if name in self._writeMethods and session not in self._sessions and request.method != "OPTIONS":
            return resource.NoResource("No such child resource.")
        # Generate session if doesn't exist, unless just checking options
        if session not in self._sessions and request.method != "OPTIONS":
            self._sessions[session] = Stub(self, session)
        # Delegate request to appropriate handler
        return self._methods[name](self, self._sessions[session] if request.method != "OPTIONS" else None)
    
    def putChild(self, path, child):
        child.parent = self
        resource.Resource.putChild(self, path, child)
    
    def setBaseHeaders(self, request, cookie=True):
        origin = request.getHeader("Origin")
        headers = request.getHeader('Access-Control-Request-Headers')
        if origin is None or origin == "null":
            origin = "*"
        request.setHeader('access-control-allow-origin', origin)
        request.setHeader('access-control-allow-credentials', 'true')
        request.setHeader('Cache-Control', 'no-store, no-cache, must-revalidate, max-age=0')
        if headers is not None:
            request.setHeader('Access-Control-Allow-Headers', headers)
        if self._options["cookie_needed"] and cookie:
            cookie = request.getCookie("JSESSIONID") if request.getCookie("JSESSIONID") else "dummy"
            request.addCookie("JSESSIONID", cookie, path="/")
    
    def render_GET(self, request):
        self.setBaseHeaders(request,False)
        request.setHeader('content-type', 'text/plain; charset=UTF-8')
        return "Welcome to SockJS!\n"
